packages/synapse-react-client/src/components/DataGrid/hooks/useGridUndoRedo.test.ts

```diff
@@ -0,0 +1,242 @@
+      const { result } = renderHook(() =>
+        useGridUndoRedo(mockOnApplyModelChange),
+      )
+      act(() => {
+        result.current.addToUndoStack({
+          type: 'CREATE' as const,
+          rowIndex: 0,
+          newValue: { id: '1' },
+        })
+        result.current.addToUndoStack({
+          type: 'CREATE' as const,
+          rowIndex: 1,
+          newValue: { id: '2' },
+        })
+        result.current.addToUndoStack({
+          type: 'UPDATE' as const,
+          rowIndex: 0,
+          previousValue: { id: '1' },
+          newValue: { id: '1-updated' },
+        })
+      })
+
+      expect(result.current.undoPreview).toEqual({
+        lastType: 'UPDATE',
+        sameTypeCount: 1,
+        totalActions: 3,
+      })
+    })
+
+    it('should remove the last action from the undo stack when handleUndo is called', () => {
+      const { result } = renderHook(() =>
+        useGridUndoRedo(mockOnApplyModelChange),
+      )
+
+      const action: GridAction = {
+        type: 'CREATE',
+        rowIndex: 0,
+        newValue: { id: '1' },
+      }
+
+      act(() => {
+        result.current.addToUndoStack(action)
+      })
+
+      // Confirm the action is in the stack
+      expect(result.current.undoPreview?.totalActions).toBe(1)
+      expect(result.current.undoPreview?.lastType).toBe('CREATE')
+
+      // Perform undo
+      act(() => {
+        result.current.handleUndo()
+      })
+
+      // The action should be removed from the stack
+      expect(result.current.undoPreview).toBeNull()
+
+      // onApplyModelChange should still have been called with the inverse
+      expect(mockOnApplyModelChange).toHaveBeenCalled()
+      // Update this line to expect both parameters
+      expect(convertActionToModelChange).toHaveBeenCalledWith(action, 'undo')
+    })
+
+    it('should undo multiple actions one by one', () => {
+      const { result } = renderHook(() =>
+        useGridUndoRedo(mockOnApplyModelChange),
+      )
+
+      const actions: GridAction[] = [
+        { type: 'CREATE', rowIndex: 0, newValue: { id: '1' } },
+        {
+          type: 'UPDATE',
+          rowIndex: 0,
+          previousValue: { id: '1' },
+          newValue: { id: '1-updated' },
+        },
+      ]
+
+      act(() => {
+        actions.forEach(action => result.current.addToUndoStack(action))
+      })
+
+      expect(result.current.undoPreview?.totalActions).toBe(2)
+
+      act(() => result.current.handleUndo())
+      expect(result.current.undoPreview?.totalActions).toBe(1)
+
+      act(() => result.current.handleUndo())
+      expect(result.current.undoPreview).toBeNull()
+
+      expect(mockOnApplyModelChange).toHaveBeenCalledTimes(2)
+    })
+  })
+
+  describe('Redo functionality', () => {
+    it('should initialize with empty redo stack', () => {
+      const { result } = renderHook(() =>
+        useGridUndoRedo(mockOnApplyModelChange),
+      )
+      expect(result.current.redoPreview).toBeNull()
```

Comment:
The UI will probably change (or be removed/decoupled from this hook) as we add more polish to the grid. It would be a bit more stable to pin our test to the data that the hook returns (testing the UI now is ok too, since it is still created by this hook!)
```suggestion
      expect(result.current.redoPreview).toBeNull()
      expect(result.current.redoTotalActions).toBe(0)
```